File: static/upload_admin.js
(() => {
  let deviations = [];
  let presets = [];
  let galleries = [];
  let selectedPresetId = null;

  function showLoading(text = "Processing...") {
    document.getElementById("loadingText").textContent = text;
    document.getElementById("loading").classList.add("active");
    document.getElementById("overlay").classList.add("active");
  }

  function hideLoading() {
    document.getElementById("loading").classList.remove("active");
    document.getElementById("overlay").classList.remove("active");
  }

  function setStatus(message, type = "info") {
    const statusEl = document.getElementById("statusMessage");
    if (!statusEl) return;
    statusEl.textContent = message;
    const color =
      type === "error" ? "#dc3545" : type === "success" ? "#198754" : "#6c757d";
    statusEl.style.color = color;
  }

  function toggleMatureOptions() {
    const isMature = document.getElementById("isMature").checked;
    document.getElementById("matureOptions").style.display = isMature ? "block" : "none";
  }

  async function loadGalleries() {
    try {
      const response = await fetch("/api/admin/galleries");
      const data = await response.json();

      if (data.success) {
        galleries = data.galleries;
        const select = document.getElementById("gallerySelect");
        select.innerHTML = '<option value="">-- No Gallery --</option>';

        galleries.forEach((gallery) => {
          const option = document.createElement("option");
          option.value = gallery.folderid;
          option.textContent = gallery.name + (gallery.size ? ` (${gallery.size})` : "");
          select.appendChild(option);
        });
      }
    } catch (error) {
      console.error("Failed to load galleries:", error);
    }
  }

  async function loadPresets() {
    try {
      const response = await fetch("/api/admin/presets");
      const data = await response.json();

      if (data.success) {
        presets = data.presets;
        const select = document.getElementById("presetSelect");
        select.innerHTML = '<option value="">-- New Preset --</option>';

        presets.forEach((preset) => {
          const option = document.createElement("option");
          option.value = preset.id;
          option.textContent = preset.name + (preset.is_default ? " (Default)" : "");
          select.appendChild(option);
        });

        const defaultPreset = presets.find((p) => p.is_default);
        if (defaultPreset) {
          select.value = defaultPreset.id;
          onPresetChange();
        }
      }
    } catch (error) {
      console.error("Failed to load presets:", error);
      setStatus("Failed to load presets", "error");
    }
  }

  function onPresetChange() {
    const select = document.getElementById("presetSelect");
    selectedPresetId = select.value ? parseInt(select.value, 10) : null;

    if (selectedPresetId) {
      const preset = presets.find((p) => p.id === selectedPresetId);
      if (preset) {
        loadPresetToForm(preset);
      }
    } else {
      clearForm();
    }
    updateButtonStates();
  }

  function loadPresetToForm(preset) {
    document.getElementById("presetName").value = preset.name || "";
    document.getElementById("baseTitle").value = preset.base_title || "";
    document.getElementById("titleIncrement").value =
      preset.last_used_increment || preset.title_increment_start || 1;
    document.getElementById("gallerySelect").value = preset.gallery_folderid || "";
    document.getElementById("tags").value = Array.isArray(preset.tags)
      ? preset.tags.join(", ")
      : preset.tags || "";
    document.getElementById("artistComments").value = preset.artist_comments || "";
    document.getElementById("displayResolution").value = preset.display_resolution || 0;
    document.getElementById("isAiGenerated").checked = preset.is_ai_generated !== false;
    document.getElementById("feature").checked = preset.feature !== false;
    document.getElementById("allowComments").checked = preset.allow_comments !== false;
    document.getElementById("allowFreeDownload").checked = preset.allow_free_download || false;
    document.getElementById("isMature").checked = preset.is_mature || false;

    if (preset.is_mature) {
      document.getElementById("matureLevel").value = preset.mature_level || "";
      const classifications = preset.mature_classification || [];
      document.getElementById("matureNudity").checked = classifications.includes("nudity");
      document.getElementById("matureSexual").checked = classifications.includes("sexual");
      document.getElementById("matureGore").checked = classifications.includes("gore");
      document.getElementById("matureLanguage").checked = classifications.includes("language");
      document.getElementById("matureIdeology").checked = classifications.includes("ideology");
    }

    toggleMatureOptions();
  }

  function clearForm() {
    document.getElementById("presetName").value = "";
    document.getElementById("baseTitle").value = "";
    document.getElementById("titleIncrement").value = 1;
    document.getElementById("gallerySelect").value = "";
    document.getElementById("tags").value = "";
    document.getElementById("artistComments").value = "";
    document.getElementById("displayResolution").value = 0;
    document.getElementById("isAiGenerated").checked = true;
    document.getElementById("feature").checked = true;
    document.getElementById("allowComments").checked = true;
    document.getElementById("allowFreeDownload").checked = false;
    document.getElementById("isMature").checked = false;
    document.getElementById("matureLevel").value = "";
    document.getElementById("matureNudity").checked = false;
    document.getElementById("matureSexual").checked = false;
    document.getElementById("matureGore").checked = false;
    document.getElementById("matureLanguage").checked = false;
    document.getElementById("matureIdeology").checked = false;
    toggleMatureOptions();
  }

  function getFormData() {
    const tags = document
      .getElementById("tags")
      .value.split(",")
      .map((t) => t.trim())
      .filter((t) => t.length > 0);

    const matureClassification = [];
    if (document.getElementById("matureNudity").checked) matureClassification.push("nudity");
    if (document.getElementById("matureSexual").checked) matureClassification.push("sexual");
    if (document.getElementById("matureGore").checked) matureClassification.push("gore");
    if (document.getElementById("matureLanguage").checked) matureClassification.push("language");
    if (document.getElementById("matureIdeology").checked) matureClassification.push("ideology");

    return {
      name: document.getElementById("presetName").value,
      base_title: document.getElementById("baseTitle").value,
      title_increment_start: parseInt(document.getElementById("titleIncrement").value, 10) || 1,
      gallery_folderid: document.getElementById("gallerySelect").value || null,
      tags,
      artist_comments: document.getElementById("artistComments").value || null,
      display_resolution: parseInt(document.getElementById("displayResolution").value, 10) || 0,
      is_ai_generated: document.getElementById("isAiGenerated").checked,
      feature: document.getElementById("feature").checked,
      allow_comments: document.getElementById("allowComments").checked,
      allow_free_download: document.getElementById("allowFreeDownload").checked,
      is_mature: document.getElementById("isMature").checked,
      mature_level: document.getElementById("matureLevel").value || null,
      mature_classification: matureClassification,
    };
  }

  async function savePreset() {
    const formData = getFormData();

    if (!formData.name || !formData.base_title) {
      setStatus("Preset Name and Base Title are required", "error");
      return;
    }

    showLoading("Saving preset...");

    try {
      const response = await fetch("/api/admin/presets", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(formData),
      });

      const data = await response.json();

      if (data.success) {
        setStatus("Preset saved successfully", "success");
        await loadPresets();
        document.getElementById("presetSelect").value = data.preset_id;
        selectedPresetId = data.preset_id;
        updateButtonStates();
      } else {
        setStatus("Save failed: " + data.error, "error");
      }
    } catch (error) {
      console.error("Save failed:", error);
      setStatus("Save failed: " + error.message, "error");
    } finally {
      hideLoading();
    }
  }

  async function loadDrafts() {
    try {
      const response = await fetch("/api/admin/drafts");
      const data = await response.json();

      if (data.success) {
        deviations = data.deviations;
        renderTable();
        setStatus(`Loaded ${deviations.length} deviations`);
      }
    } catch (error) {
      console.error("Failed to load drafts:", error);
      setStatus("Failed to load drafts", "error");
    }
  }

  async function scanFiles() {
    showLoading("Scanning upload folder...");

    try {
      const response = await fetch("/api/admin/scan", { method: "POST" });
      const data = await response.json();

      if (data.success) {
        deviations = data.drafts;
        renderTable();
        setStatus(`Scanned: ${data.count} files`, "success");
      } else {
        setStatus("Scan failed: " + data.error, "error");
      }
    } catch (error) {
      console.error("Scan failed:", error);
      setStatus("Scan failed: " + error.message, "error");
    } finally {
      hideLoading();
    }
  }

  function statusBadge(status) {
    const normalized = (status || "").toLowerCase();
    const cls =
      normalized === "stashing"
        ? "status-stashing"
        : normalized === "stashed"
        ? "status-stashed"
        : normalized === "publishing"
        ? "status-publishing"
        : normalized === "published"
        ? "status-published"
        : normalized === "failed"
        ? "status-failed"
        : "status-draft";
    return `<span class="badge status-badge ${cls}">${status || "draft"}</span>`;
  }

  function renderTable() {
    const tbody = document.getElementById("tableBody");

    if (!deviations.length) {
      tbody.innerHTML = `
        <tr>
          <td colspan="6" class="text-center text-muted py-4">
            <div class="fw-semibold mb-1">No files found</div>
            <div class="small">Upload images to the upload folder and click "Scan Files"</div>
          </td>
        </tr>
      `;
      updateButtonStates();
      return;
    }

    tbody.innerHTML = deviations
      .map(
        (dev) => `
          <tr>
            <td><input type="checkbox" class="form-check-input deviation-checkbox" data-id="${dev.id}" onchange="updateButtonStates()"></td>
            <td>
              <img src="/api/admin/thumbnail/${dev.id}" alt="${dev.filename}" class="thumbnail"
                   onerror="this.src='data:image/svg+xml,%3Csvg xmlns=%22http://www.w3.org/2000/svg%22 width=%2264%22 height=%2264%22%3E%3Crect fill=%22%23e9ecef%22 width=%2264%22 height=%2264%22/%3E%3Ctext x=%2250%25%22 y=%2250%25%22 text-anchor=%22middle%22 dy=%22.3em%22 fill=%22%23999%22 font-size=%228%22%3ENo Image%3C/text%3E%3C/svg%3E'">
            </td>
            <td class="text-nowrap">${dev.filename}</td>
            <td>${dev.title || "-"}</td>
            <td>${statusBadge(dev.status)}</td>
            <td>
              <button class="btn btn-outline-danger btn-sm" onclick="deleteSingle(${dev.id})">Delete</button>
            </td>
          </tr>
        `
      )
      .join("");

    updateButtonStates();
  }

  function toggleSelectAll() {
    const selectAll = document.getElementById("selectAll");
    document.querySelectorAll(".deviation-checkbox").forEach((cb) => {
      cb.checked = selectAll.checked;
    });
    updateButtonStates();
  }

  function updateButtonStates() {
    const selected = getSelectedIds();
    const hasSelection = selected.length > 0;
    const hasPresetConfig =
      document.getElementById("presetName").value && document.getElementById("baseTitle").value;

    document.getElementById("uploadBtn").disabled = !hasSelection || !hasPresetConfig;
    document.getElementById("deleteBtn").disabled = !hasSelection;
  }

  function getSelectedIds() {
    return Array.from(document.querySelectorAll(".deviation-checkbox:checked")).map((cb) =>
      parseInt(cb.dataset.id, 10)
    );
  }

  async function uploadSelected() {
    const selectedIds = getSelectedIds();
    if (!selectedIds.length) return;

    const formData = getFormData();
    if (!formData.name || !formData.base_title) {
      setStatus("Please fill in Preset Name and Base Title", "error");
      return;
    }

    showLoading(`Uploading ${selectedIds.length} files to DeviantArt...`);

    try {
      const presetResponse = await fetch("/api/admin/presets", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(formData),
      });
      const presetData = await presetResponse.json();

      if (!presetData.success) {
        setStatus("Failed to save preset: " + presetData.error, "error");
        hideLoading();
        return;
      }

      const presetId = presetData.preset_id;

      const response = await fetch("/api/admin/upload", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          preset_id: presetId,
          deviation_ids: selectedIds,
        }),
      });

      const data = await response.json();

      if (data.success) {
        const results = data.results || {};
        const successCount = results.success ? results.success.length : 0;
        const failedCount = results.failed ? results.failed.length : 0;

        await loadDrafts();
        await loadPresets();

        if (failedCount === 0) {
          setStatus(`Successfully uploaded ${successCount} files!`, "success");
        } else {
          setStatus(`Uploaded ${successCount}, failed ${failedCount}`, failedCount > 0 ? "error" : "success");
        }
      } else {
        setStatus("Upload failed: " + data.error, "error");
      }
    } catch (error) {
      console.error("Upload failed:", error);
      setStatus("Upload failed: " + error.message, "error");
    } finally {
      hideLoading();
    }
  }

  async function deleteSelected() {
    const selectedIds = getSelectedIds();
    if (!selectedIds.length) return;

    if (!confirm(`Delete ${selectedIds.length} selected files and their database records?`)) return;

    showLoading(`Deleting ${selectedIds.length} files...`);

    try {
      const response = await fetch("/api/admin/delete", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ deviation_ids: selectedIds }),
      });

      const data = await response.json();

      if (data.success) {
        await loadDrafts();
        setStatus(`Deleted ${data.count} files`, "success");
      } else {
        setStatus("Delete failed: " + data.error, "error");
      }
    } catch (error) {
      console.error("Delete failed:", error);
      setStatus("Delete failed: " + error.message, "error");
    } finally {
      hideLoading();
    }
  }

  async function deleteSingle(id) {
    if (!confirm("Delete this file and its database record?")) return;

    showLoading("Deleting...");

    try {
      const response = await fetch("/api/admin/delete", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ deviation_ids: [id] }),
      });

      const data = await response.json();

      if (data.success) {
        await loadDrafts();
        setStatus("File deleted", "success");
      } else {
        setStatus("Delete failed: " + data.error, "error");
      }
    } catch (error) {
      console.error("Delete failed:", error);
      setStatus("Delete failed: " + error.message, "error");
    } finally {
      hideLoading();
    }
  }

  document.addEventListener("DOMContentLoaded", () => {
    loadGalleries();
    loadPresets();
    loadDrafts();
  });

  // Expose for inline handlers
  window.onPresetChange = onPresetChange;
  window.toggleMatureOptions = toggleMatureOptions;
  window.savePreset = savePreset;
  window.scanFiles = scanFiles;
  window.uploadSelected = uploadSelected;
  window.deleteSelected = deleteSelected;
  window.deleteSingle = deleteSingle;
  window.toggleSelectAll = toggleSelectAll;
  window.updateButtonStates = updateButtonStates;
})();

File: src/api/stats_routes/upload_admin.py
"""Upload admin API routes."""

from __future__ import annotations

from collections.abc import Callable

from flask import Flask, g, jsonify, request

from ...domain.models import UploadPreset


def register_upload_admin_routes(
    app: Flask,
    *,
    get_upload_services: Callable[[], tuple[object, object, object]],
    get_repositories: Callable[[], tuple[object, ...]],
) -> None:
    """Register upload admin endpoints."""

    @app.route("/api/admin/scan", methods=["POST"])
    def scan_files():
        """Scan upload folder and create draft deviations."""
        try:
            uploader_service, _preset_repo, _deviation_repo = get_upload_services()
            drafts = uploader_service.scan_and_create_drafts()

            result = []
            for draft in drafts:
                result.append(
                    {
                        "id": draft.deviation_id,
                        "filename": draft.filename,
                        "title": draft.title,
                        "file_path": draft.file_path,
                        "status": (
                            draft.status.value
                            if hasattr(draft.status, "value")
                            else draft.status
                        ),
                        "itemid": draft.itemid,
                        "deviationid": draft.deviationid,
                        "url": draft.url,
                    }
                )

            return jsonify({"success": True, "drafts": result, "count": len(result)})
        except Exception as e:  # noqa: BLE001
            g.logger.error(f"Scan failed: {e}", exc_info=True)
            return jsonify({"success": False, "error": str(e)}), 500

    @app.route("/api/admin/drafts", methods=["GET"])
    def get_drafts():
        """Get all draft deviations from database."""
        try:
            _uploader_service, _preset_repo, deviation_repo = get_upload_services()

            all_deviations = deviation_repo.get_all_deviations()

            result = []
            for dev in all_deviations:
                result.append(
                    {
                        "id": dev.deviation_id,
                        "filename": dev.filename,
                        "title": dev.title,
                        "file_path": dev.file_path,
                        "status": (
                            dev.status.value if hasattr(dev.status, "value") else dev.status
                        ),
                        "itemid": dev.itemid,
                        "deviationid": dev.deviationid,
                        "url": dev.url,
                        "error": dev.error,
                        "tags": dev.tags,
                        "is_mature": dev.is_mature,
                    }
                )

            return jsonify(
                {"success": True, "deviations": result, "count": len(result)}
            )
        except Exception as e:  # noqa: BLE001
            g.logger.error(f"Get drafts failed: {e}", exc_info=True)
            return jsonify({"success": False, "error": str(e)}), 500

    @app.route("/api/admin/galleries", methods=["GET"])
    def get_galleries_for_admin():
        """Get all galleries for dropdown selection."""
        try:
            (
                _user_repo,
                _token_repo,
                gallery_repo,
                _deviation_repo,
                _deviation_stats_repo,
                _stats_snapshot_repo,
                _user_stats_snapshot_repo,
                _deviation_metadata_repo,
            ) = get_repositories()
            galleries = gallery_repo.get_all_galleries()

            result = []
            for gallery in galleries:
                result.append(
                    {
                        "id": gallery.gallery_db_id,
                        "folderid": gallery.folderid,
                        "name": gallery.name,
                        "size": gallery.size,
                        "parent": gallery.parent,
                    }
                )

            return jsonify({"success": True, "galleries": result, "count": len(result)})
        except Exception as e:  # noqa: BLE001
            g.logger.error(f"Get galleries failed: {e}", exc_info=True)
            return jsonify({"success": False, "error": str(e)}), 500

    @app.route("/api/admin/presets", methods=["GET"])
    def get_presets():
        """Get all presets for dropdown."""
        try:
            _uploader_service, preset_repo, _deviation_repo = get_upload_services()
            presets = preset_repo.get_all_presets()

            result = []
            for preset in presets:
                result.append(
                    {
                        "id": preset.preset_id,
                        "name": preset.name,
                        "description": preset.description,
                        "base_title": preset.base_title,
                        "title_increment_start": preset.title_increment_start,
                        "last_used_increment": preset.last_used_increment,
                        "is_default": preset.is_default,
                        "tags": preset.tags,
                        "is_mature": preset.is_mature,
                        "gallery_folderid": preset.gallery_folderid,
                    }
                )

            return jsonify({"success": True, "presets": result, "count": len(result)})
        except Exception as e:  # noqa: BLE001
            g.logger.error(f"Get presets failed: {e}", exc_info=True)
            return jsonify({"success": False, "error": str(e)}), 500

    @app.route("/api/admin/presets", methods=["POST"])
    def save_preset():
        """Save or update a preset."""
        try:
            data = request.get_json()

            if not data.get("name") or not data.get("base_title"):
                return (
                    jsonify(
                        {
                            "success": False,
                            "error": "Name and base_title are required",
                        }
                    ),
                    400,
                )

            _uploader_service, preset_repo, _deviation_repo = get_upload_services()

            preset = UploadPreset(
                name=data["name"],
                description=data.get("description"),
                base_title=data["base_title"],
                title_increment_start=data.get("title_increment_start", 1),
                last_used_increment=data.get("last_used_increment", 1),
                artist_comments=data.get("artist_comments"),
                tags=data.get("tags", []),
                is_ai_generated=data.get("is_ai_generated", True),
                noai=data.get("noai", False),
                is_dirty=data.get("is_dirty", False),
                is_mature=data.get("is_mature", False),
                mature_level=data.get("mature_level"),
                mature_classification=data.get("mature_classification", []),
                feature=data.get("feature", True),
                allow_comments=data.get("allow_comments", True),
                display_resolution=data.get("display_resolution", 0),
                allow_free_download=data.get("allow_free_download", False),
                add_watermark=data.get("add_watermark", False),
                gallery_folderid=data.get("gallery_folderid"),
                is_default=data.get("is_default", False),
            )

            preset_id = preset_repo.save_preset(preset)

            return jsonify(
                {
                    "success": True,
                    "preset_id": preset_id,
                    "message": "Preset saved successfully",
                }
            )
        except Exception as e:  # noqa: BLE001
            g.logger.error(f"Save preset failed: {e}", exc_info=True)
            return jsonify({"success": False, "error": str(e)}), 500

    @app.route("/api/admin/apply-preset", methods=["POST"])
    def apply_preset():
        """Apply preset to selected deviations."""
        try:
            data = request.get_json()
            preset_id = data.get("preset_id")
            deviation_ids = data.get("deviation_ids", [])

            if not preset_id or not deviation_ids:
                return (
                    jsonify(
                        {
                            "success": False,
                            "error": "preset_id and deviation_ids required",
                        }
                    ),
                    400,
                )

            uploader_service, preset_repo, deviation_repo = get_upload_services()

            preset = preset_repo.get_preset_by_id(preset_id)
            if not preset:
                return jsonify({"success": False, "error": "Preset not found"}), 404

            applied = []
            for dev_id in deviation_ids:
                deviation = deviation_repo.get_deviation_by_id(dev_id)
                if deviation:
                    increment = preset_repo.increment_preset_counter(preset_id)
                    uploader_service.apply_preset_to_deviation(deviation, preset, increment)
                    deviation_repo.update_deviation(deviation)
                    applied.append(dev_id)

            return jsonify({"success": True, "applied": applied, "count": len(applied)})
        except Exception as e:  # noqa: BLE001
            g.logger.error(f"Apply preset failed: {e}", exc_info=True)
            return jsonify({"success": False, "error": str(e)}), 500

    @app.route("/api/admin/stash", methods=["POST"])
    def stash_selected():
        """Stash selected deviations."""
        try:
            data = request.get_json()
            deviation_ids = data.get("deviation_ids", [])
            preset_id = data.get("preset_id")

            if not deviation_ids or not preset_id:
                return (
                    jsonify(
                        {
                            "success": False,
                            "error": "deviation_ids and preset_id required",
                        }
                    ),
                    400,
                )

            uploader_service, preset_repo, _deviation_repo = get_upload_services()

            preset = preset_repo.get_preset_by_id(preset_id)
            if not preset:
                return jsonify({"success": False, "error": "Preset not found"}), 404

            results = uploader_service.batch_stash(deviation_ids, preset)

            return jsonify({"success": True, "results": results})
        except Exception as e:  # noqa: BLE001
            g.logger.error(f"Stash failed: {e}", exc_info=True)
            return jsonify({"success": False, "error": str(e)}), 500

    @app.route("/api/admin/publish", methods=["POST"])
    def publish_selected():
        """Publish stashed deviations."""
        try:
            data = request.get_json()
            deviation_ids = data.get("deviation_ids", [])

            if not deviation_ids:
                return jsonify({"success": False, "error": "deviation_ids required"}), 400

            uploader_service, _preset_repo, _deviation_repo = get_upload_services()

            results = uploader_service.batch_publish(deviation_ids)

            return jsonify({"success": True, "results": results})
        except Exception as e:  # noqa: BLE001
            g.logger.error(f"Publish failed: {e}", exc_info=True)
            return jsonify({"success": False, "error": str(e)}), 500

    @app.route("/api/admin/upload", methods=["POST"])
    def upload_selected():
        """Upload selected deviations (stash + publish)."""
        try:
            data = request.get_json()
            deviation_ids = data.get("deviation_ids", [])
            preset_id = data.get("preset_id")

            if not deviation_ids or not preset_id:
                return (
                    jsonify(
                        {
                            "success": False,
                            "error": "deviation_ids and preset_id required",
                        }
                    ),
                    400,
                )

            uploader_service, preset_repo, _deviation_repo = get_upload_services()

            preset = preset_repo.get_preset_by_id(preset_id)
            if not preset:
                return jsonify({"success": False, "error": "Preset not found"}), 404

            results = uploader_service.batch_upload(deviation_ids, preset)

            return jsonify({"success": True, "results": results})
        except Exception as e:  # noqa: BLE001
            g.logger.error(f"Upload failed: {e}", exc_info=True)
            return jsonify({"success": False, "error": str(e)}), 500

    @app.route("/api/admin/delete", methods=["POST"])
    def delete_selected():
        """Delete deviations and files."""
        try:
            data = request.get_json()
            deviation_ids = data.get("deviation_ids", [])

            if not deviation_ids:
                return jsonify({"success": False, "error": "deviation_ids required"}), 400

            uploader_service, _preset_repo, _deviation_repo = get_upload_services()

            deleted = []
            failed = []

            for dev_id in deviation_ids:
                if uploader_service.delete_deviation_and_file(dev_id):
                    deleted.append(dev_id)
                else:
                    failed.append(dev_id)

            return jsonify(
                {"success": True, "deleted": deleted, "failed": failed, "count": len(deleted)}
            )
        except Exception as e:  # noqa: BLE001
            g.logger.error(f"Delete failed: {e}", exc_info=True)
            return jsonify({"success": False, "error": str(e)}), 500
